augur/io/metadata.py

```diff
@@ -88,3 +97,228 @@ def read_metadata(metadata_file, id_columns=("strain", "name"), chunk_size=None)
+            yield from metadata_reader
+        else:
+            if id_column is None:
+                id_column = metadata_reader.fieldnames[0]
+
+            for record in metadata_reader:
+                record_id = record.get(id_column)
+                if record_id is None:
+                    raise AugurError(f"The provided id column {id_column!r} does not exist in the metadata.")
+
+                if record_id in seen_ids:
+                    duplicate_ids.add(record_id)
+                else:
+                    seen_ids.add(record_id)
+
+                yield record
+
+    if duplicate_reporting is not DataErrorMethod.SILENT and duplicate_ids:
+        duplicates_message = f"The following records are duplicated in {metadata!r}:\n" + "\n".join(sorted(duplicate_ids))
+
+        if duplicate_reporting is DataErrorMethod.WARN:
+            print_err(f"WARNING: {duplicates_message}")
+        elif duplicate_reporting is DataErrorMethod.ERROR:
+            raise AugurError(duplicates_message)
+        else:
+            raise AugurError(f"Encountered unhandled duplicate reporting method: {duplicate_reporting!r}")
+
+
+def read_metadata_with_sequences(metadata, fasta, seq_id_column, seq_field='sequence',
+    unmatched_reporting=DataErrorMethod.ERROR, duplicate_reporting=DataErrorMethod.ERROR):
+    """
+    Read rows from *metadata* file and yield each row as a single dict that has
+    been updated with their corresponding sequence from the *fasta* file.
+    Matches the metadata record with sequences using the sequence id provided
+    in the *seq_id_column*. To ensure that the sequences can be matched with
+    the metadata, the FASTA headers should only contain the sequence id.
+
+    Will report unmatched records if requested via *unmatched_reporting* after
+    the generator has been exhausted.
+
+    Will report duplicate records if requested via *duplicate_reporting* after
+    the generator has been exhausted.
+
+    Reads the *fasta* file with `pyfastx.Fasta`, which creates an index for
+    the file to allow random access of sequences via the sequence id.
+    See pyfastx docs for more details:
+    https://pyfastx.readthedocs.io/en/latest/usage.html#fasta
+
+    Parameters
+    ----------
+    metadata: str
+        Path to a CSV or TSV metadata file
+
+    fasta: str
+        Path to a plain or gzipped FASTA file
+
+    seq_id_column: str
+        The column in the metadata file that contains the sequence id for
+        matching sequences
+
+    seq_field: str, optional
+        The field name to use for the sequence in the updated record
+
+    unmatched_reporting: DataErrorMethod, optional
+        How should unmatched records be reported
+
+    duplicate_reporting: DataErrorMethod, optional
+        How should duplicate records be reported
+
+    Yields
+    ------
+    dict
+        The parsed metadata record with the sequence
+    """
+    sequences = pyfastx.Fasta(fasta)
+    sequence_ids = set(sequences.keys())
```

Comment:
Reading the entire FASTA for each `augur curate` subcommand won't scale well for large datasets. But I can't think of a good solution here other than using databases (#866).

Reply:
See continued conversation around reading set of ids in https://github.com/nextstrain/augur/pull/1039#discussion_r975879350